### ISSUE TITLE
missing file

### ISSUE THREAD

**Issue by GFTwrt** [nitadori/Hermite#1]
Hello,

try to compile your code. But got problems with gcc on Ubuntu 64 System I had no success to compile. Than tried to compile with CUDA. but the file helpers_cuda.h is missing.

In file included from ../SRC/hermite8-titan.h:2,
                 from ../SRC/hermite8.h:313,
                 from ../SRC/hermite8-titan.cu:16:
../SRC/cuda_pointer.h:8:12: fatal error: helper_cuda.h: missing file
    8 | #  include <helper_cuda.h>

Can you help me?

**Comment by nitadori** [comment]
This file is provided in cuda-samples.
https://github.com/NVIDIA/cuda-samples/blob/master/Common/helper_cuda.h

**Comment by GFTwrt** [comment]
Thank You!
In the meanwhile I had seen that this is just debugging code I commented out the include as well as the checkCudaErrors  in the following line.
But while searching for the file I found something other interesting in your repository: The folder "HighOrder". Are this higher order codes working well? I'm searching for a code up to 20 particles with  dE/E of 10E-30 or better according the scheme in your paper  "6th and 8th Order Hermite Integrator for N -body Simulations".
Why do you use dd_real? My investigations shows that mpfr is faster on my CPU (AMD Ryzen 7) by usage of the same precision. Have you ever tied CAMPARY https://homepages.laas.fr/mmjoldes/campary/ ? it is binary cast able to and from dd_real but does not help me because of the low amount of particles which calculates faster at the CPU.

Thank You for your support! 
Thomas

**Comment by nitadori** [comment]
That repository is up to 16th-order implementations of Hermite integrators.
I have only tested for Kepler and ∞-shaped 3-body systems.
"dd_real" is about 10x slower than the native double, but I have never imagined mpfr is faster.

**Comment by GFTwrt** [comment]
Here the output of eight-16th:
...
62.832031 -1.110841e-30  0.647488 -0.357915  -1.077439 0.046748  0.429951 0.311167
e1 :      -1.2871419917663256
9.765625e-04 **1.235332e-30** 10 1

I like the "**e-30**" in this output.
Your sources do not contain any copyright or copyleft. Would it be possible that you would insert appropriate comments in the head?  I*d like to add the engine/code to https://github.com/amusecode/amuse to use it afterwards for some calculations. The mpfr code called "Brutus" is only 2nd order and therefore to slow.

Additionally I'd like to add one additional force per particle valid while one dt to interact with a hydro code. I'd add it to the calculation of x" before doing higher deviations. Do you see any problems with this?

If my calculations shows the expected is there any paper describing your code or should I just refer to "6th and 8th Order Hermite Integrator for N -body Simulations"?

Best regards
Thomas

**Comment by nitadori** [comment]
I have added a LICENSE file (MIT) there.
You can freely modify the code.
I think you can also cite Dittmann (2020).
  https://arxiv.org/abs/2006.02501

**Comment by GFTwrt** [comment]
Thank You!

In the meanwhile I checked your code. It is written really clear and easy to read.

The deeper analysis showed that you use dd_real. At least for the 16th-order code this seams not be enough. It should use qd_real. dE/E is at the theoretical limit of dd_real (two double containing 15.5 significant each).

**Comment by nitadori** [comment]
You are right, it reaches the limit of dd from dt=1/64, for T=2π ∞-shaped three-body problem.

![image](https://user-images.githubusercontent.com/4830914/111074694-836b5400-8527-11eb-98d6-33721560fbee.png)
